From 334a0ce52cfee7464809b21d770c8ac8767d393d Mon Sep 17 00:00:00 2001
From: Chris <meischris39@hotmail.com>
Date: Thu, 22 Oct 2026 15:54:42 +0100
Subject: [PATCH] [RetroPlayer] Give savestate thumbnails the game's aspect
 ratio

Thumbnails were scaled from the frame's pixels as if they were square,
so a game drawn with non-square pixels, such as the SNES at 256x224
shown at 4:3, looked squashed next to how it plays. Scale to the shape
the frame is shown at in the Normal stretch mode instead.
---
 .../xml/Includes_DialogSelect.xml             |  4 +--
 .../RetroPlayer/rendering/RPRenderManager.cpp | 19 ++++++++---
 .../RetroPlayer/rendering/RenderUtils.cpp     | 11 ++++++
 .../cores/RetroPlayer/rendering/RenderUtils.h |  9 +++++
 .../RetroPlayer/rendering/test/CMakeLists.txt |  1 +
 .../rendering/test/TestRenderUtils.cpp        | 34 +++++++++++++++++++
 6 files changed, 71 insertions(+), 7 deletions(-)
 create mode 100644 xbmc/cores/RetroPlayer/rendering/test/TestRenderUtils.cpp

diff --git a/addons/skin.estuary/xml/Includes_DialogSelect.xml b/addons/skin.estuary/xml/Includes_DialogSelect.xml
index ee7d629..e1f5139 100644
--- a/addons/skin.estuary/xml/Includes_DialogSelect.xml
+++ b/addons/skin.estuary/xml/Includes_DialogSelect.xml
@@ -813,7 +813,7 @@
 								<left>10</left>
 								<right>10</right>
 								<top>10</top>
-								<height>256</height>
+								<height>250</height>
 								<texture>$INFO[ListItem.Art(screenshot)]</texture>
 								<aspectratio>keep</aspectratio>
 							</control>
@@ -851,7 +851,7 @@
 								<left>10</left>
 								<right>10</right>
 								<top>10</top>
-								<height>256</height>
+								<height>250</height>
 								<texture>$INFO[ListItem.Art(screenshot)]</texture>
 								<aspectratio>keep</aspectratio>
 							</control>
diff --git a/xbmc/cores/RetroPlayer/rendering/RPRenderManager.cpp b/xbmc/cores/RetroPlayer/rendering/RPRenderManager.cpp
index 59bc8fd..481f981 100644
--- a/xbmc/cores/RetroPlayer/rendering/RPRenderManager.cpp
+++ b/xbmc/cores/RetroPlayer/rendering/RPRenderManager.cpp
@@ -11,6 +11,7 @@
 #include "RenderContext.h"
 #include "RenderSettings.h"
 #include "RenderTranslator.h"
+#include "RenderUtils.h"
 #include "ServiceBroker.h"
 #include "URL.h"
 #include "cores/RetroPlayer/buffers/IRenderBuffer.h"
@@ -977,7 +978,10 @@ void CRPRenderManager::CreateRenderBuffer(IRenderBufferPool* bufferPool)
     IRenderBuffer* renderBuffer =
         CreateFromCache(m_cachedFrame, m_cachedWidth, m_cachedHeight, bufferPool, m_bufferMutex);
     if (renderBuffer != nullptr)
+    {
+      renderBuffer->SetDisplayAspectRatio(m_cachedDisplayAspectRatio);
       m_renderBuffers.emplace_back(renderBuffer);
+    }
   }
 }
 
@@ -1113,6 +1117,7 @@ void CRPRenderManager::SaveThumbnail(const std::string& thumbnailPath)
   size_t sourceSize = 0;
   unsigned int width = 0;
   unsigned int height = 0;
+  float displayAspectRatio = 0.0f;
   unsigned int rotationCCW = 0;
 
   if (renderBuffer != nullptr)
@@ -1122,15 +1127,17 @@ void CRPRenderManager::SaveThumbnail(const std::string& thumbnailPath)
     sourceSize = renderBuffer->GetFrameSize();
     width = renderBuffer->GetWidth();
     height = renderBuffer->GetHeight();
+    displayAspectRatio = renderBuffer->GetDisplayAspectRatio();
     rotationCCW = renderBuffer->GetRotation();
   }
   else if (!cachedFrame.empty())
   {
     sourceFormat = m_format;
-    sourceData = m_cachedFrame.data();
-    sourceSize = m_cachedFrame.size();
+    sourceData = cachedFrame.data();
+    sourceSize = cachedFrame.size();
     width = m_cachedWidth;
     height = m_cachedHeight;
+    displayAspectRatio = m_cachedDisplayAspectRatio;
     rotationCCW = m_cachedRotationCCW;
   }
 
@@ -1145,9 +1152,11 @@ void CRPRenderManager::SaveThumbnail(const std::string& thumbnailPath)
 
   const int stride = CRenderTranslator::TranslateWidthToBytes(width, sourceFormat);
 
-  unsigned int scaleWidth = 400;
-  unsigned int scaleHeight = 220;
-  CPicture::GetScale(width, height, scaleWidth, scaleHeight);
+  // Scale to the shape the frame is shown at, so non-square pixels aren't stretched
+  unsigned int scaleWidth = 444;
+  unsigned int scaleHeight = 250;
+  CPicture::GetScale(CRenderUtils::GetDisplayWidth(width, height, displayAspectRatio), height,
+                     scaleWidth, scaleHeight);
 
   const int bytesPerPixel = 4;
   std::vector<uint8_t> scaledImage(scaleWidth * scaleHeight * bytesPerPixel);
diff --git a/xbmc/cores/RetroPlayer/rendering/RenderUtils.cpp b/xbmc/cores/RetroPlayer/rendering/RenderUtils.cpp
index 72ebbf1..34418a6 100644
--- a/xbmc/cores/RetroPlayer/rendering/RenderUtils.cpp
+++ b/xbmc/cores/RetroPlayer/rendering/RenderUtils.cpp
@@ -10,6 +10,7 @@
 
 #include "utils/MathUtils.h"
 
+#include <algorithm>
 #include <cmath>
 
 using namespace KODI;
@@ -402,3 +403,13 @@ ViewportCoordinates CRenderUtils::ReorderDrawPoints(const CRect& destRect,
 
   return rotatedDestCoords;
 }
+
+unsigned int CRenderUtils::GetDisplayWidth(unsigned int width,
+                                           unsigned int height,
+                                           float displayAspectRatio)
+{
+  if (displayAspectRatio <= 0.0f)
+    return width;
+
+  return static_cast<unsigned int>(std::max(1L, std::lround(height * displayAspectRatio)));
+}
diff --git a/xbmc/cores/RetroPlayer/rendering/RenderUtils.h b/xbmc/cores/RetroPlayer/rendering/RenderUtils.h
index c10dc16..b0594a4 100644
--- a/xbmc/cores/RetroPlayer/rendering/RenderUtils.h
+++ b/xbmc/cores/RetroPlayer/rendering/RenderUtils.h
@@ -45,6 +45,15 @@ public:
 
   static ViewportCoordinates ReorderDrawPoints(const CRect& destRect,
                                                unsigned int orientationDegCCW);
+
+  /*!
+   * \brief Get the width a frame is shown at for its height
+   *
+   * \param displayAspectRatio The frame's display aspect ratio, or 0 for square pixels
+   */
+  static unsigned int GetDisplayWidth(unsigned int width,
+                                      unsigned int height,
+                                      float displayAspectRatio);
 };
 } // namespace RETRO
 } // namespace KODI
diff --git a/xbmc/cores/RetroPlayer/rendering/test/CMakeLists.txt b/xbmc/cores/RetroPlayer/rendering/test/CMakeLists.txt
index 4731107..aa91184 100644
--- a/xbmc/cores/RetroPlayer/rendering/test/CMakeLists.txt
+++ b/xbmc/cores/RetroPlayer/rendering/test/CMakeLists.txt
@@ -2,6 +2,7 @@ set(SOURCES TestBaseRenderBufferPool.cpp
             TestHwRenderingContextEGL.cpp
             TestRenderBufferOpenGLES.cpp
             TestRenderBufferPoolFBO.cpp
+            TestRenderUtils.cpp
             TestRPProcessInfoEGL.cpp
             TestRPRenderManager.cpp
 )
diff --git a/xbmc/cores/RetroPlayer/rendering/test/TestRenderUtils.cpp b/xbmc/cores/RetroPlayer/rendering/test/TestRenderUtils.cpp
new file mode 100644
index 0000000..c599270
--- /dev/null
+++ b/xbmc/cores/RetroPlayer/rendering/test/TestRenderUtils.cpp
@@ -0,0 +1,34 @@
+/*
+ *  Copyright (C) 2026 Team Kodi
+ *  This file is part of Kodi - https://kodi.tv
+ *
+ *  SPDX-License-Identifier: GPL-2.0-or-later
+ *  See LICENSES/README.md for more information.
+ */
+
+#include "cores/RetroPlayer/rendering/RenderUtils.h"
+
+#include <gtest/gtest.h>
+
+using namespace KODI;
+using namespace RETRO;
+
+TEST(TestRenderUtils, DisplayWidthKeepsSquarePixels)
+{
+  EXPECT_EQ(320u, CRenderUtils::GetDisplayWidth(320, 240, 0.0f));
+  EXPECT_EQ(320u, CRenderUtils::GetDisplayWidth(320, 240, -1.0f));
+}
+
+TEST(TestRenderUtils, DisplayWidthFollowsAspectRatio)
+{
+  // SNES: 256x224 shown at 4:3
+  EXPECT_EQ(299u, CRenderUtils::GetDisplayWidth(256, 224, 4.0f / 3.0f));
+
+  // A wide frame squeezed into fewer pixels
+  EXPECT_EQ(640u, CRenderUtils::GetDisplayWidth(320, 360, 16.0f / 9.0f));
+}
+
+TEST(TestRenderUtils, DisplayWidthIsNeverZero)
+{
+  EXPECT_EQ(1u, CRenderUtils::GetDisplayWidth(4, 1, 0.1f));
+}
-- 
2.53.0

